Add some reduction forward AD formulas

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69956
* #69884
* #69727
* #69740
* #69690
* __->__ #69661

Differential Revision: [D33020601](https://our.internmc.facebook.com/intern/diff/D33020601)

soulitzer committed:
Add some reduction forward AD formulas

[ghstack-poisoned]

soulitzer committed:
Update on "Add some reduction forward AD formulas"




[ghstack-poisoned]

soulitzer committed:
Update on "Add some reduction forward AD formulas"




[ghstack-poisoned]

soulitzer committed:
Update on "Add some reduction forward AD formulas"




[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1179,9 +1179,11 @@
 
 - name: prod(Tensor self, *, ScalarType? dtype=None) -> Tensor
   self: prod_backward(grad, self.to(grad.scalar_type()), result)
+  result: (prod_backward(ones_like(result), self_p.to(result.scalar_type()), result) * self_t.conj()).sum().conj()
```
could be one one and expand it.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -983,6 +985,29 @@ Tensor var_backward(Tensor grad, const Tensor& self, c10::optional<IntArrayRef>
+  auto correction = correction_opt.value_or(1);
+  if (self_p.dim() == 0 || !dim_opt.has_value()) {
+
+    return var_backward(self_t.conj(), self_p, correction).sum(all_dims_less_than(self_t.dim()), keepdim).conj();
```
maybe view_as(result) or get the result.sizes() as input to this function?

albanD reviewed (approved):
looks good if CI is happy

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

soulitzer committed:
Update on "Add some reduction forward AD formulas"


Differential Revision: [D33020601](https://our.internmc.facebook.com/intern/diff/D33020601)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

soulitzer committed:
Update on "Add some reduction forward AD formulas"


Differential Revision: [D33020601](https://our.internmc.facebook.com/intern/diff/D33020601)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

soulitzer committed:
Update on "Add some reduction forward AD formulas"


Differential Revision: [D33020601](https://our.internmc.facebook.com/intern/diff/D33020601)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

soulitzer committed:
Update on "Add some reduction forward AD formulas"


Differential Revision: [D33020601](https://our.internmc.facebook.com/intern/diff/D33020601)

[ghstack-poisoned]

soulitzer committed:
Update on "Add some reduction forward AD formulas"


Differential Revision: [D33020601](https://our.internmc.facebook.com/intern/diff/D33020601)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).

soulitzer committed:
Update on "Add some reduction forward AD formulas"


Differential Revision: [D33020601](https://our.internmc.facebook.com/intern/diff/D33020601)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33020601).
